src/models/SAFT/PCSAFT/variants/PPCSAFT.jl

```diff
@@ -143,20 +300,67 @@ function a_3(model ::PPCSAFTModel, V, T, z, _data=@f(data))
     return _a_3
 end
 
-function J2(model, V, T, z, i, j, η = @f(ζ,3),m = model.params.segment.values,ϵ = model.params.epsilon.values)
-    corr_a = PPCSAFTconsts.corr_a
-    corr_b = PPCSAFTconsts.corr_b
-    m̄ = min(sqrt(m[i]*m[j]),2*one(η))
+function a_3_dq(model ::PPCSAFTModel, V, T, z, _data=@f(data))
+    dp_comps, qp_comps = @f(polar_comps)
+    μ̄² = model.params.dipole2.values
+    Q̄² = model.params.quadrupole2.values
+    _,_,_,_,η,_ = _data
+    ∑z = sum(z)
+    ρ = N_A*∑z/V
+    _a_3 = zero(T+V+first(z))
+    m = model.params.segment.values
+    ϵ = model.params.epsilon.values
+    σ = model.params.sigma.values
+    @inbounds for i ∈ dp_comps
+        for j ∈ union(dp_comps, qp_comps)
+            for k ∈ qp_comps
+                _J3_ijk = @f(J3,type,i,j,k,η,m)
+                _a_3 += z[i]*z[j]*z[k]*σ[i,i]/
+                    (σ[k,k]*(σ[i,j]*σ[i,k]*σ[j,k])^2)*
+                    μ̄²[i]*Q̄²[k]*(σ[j,j]*μ̄²[j]+1.19374/σ[j,j]*Q̄²[j])*_J3_ijk
+            end
+        end
+    end
+    _a_3 *= -ρ^2/(T*T*T)/(∑z*∑z*∑z)
+    return _a_3
+end
+
+# Returns [Dipole Comp idxs], [Quadrupole Comp idxs]
+function polar_comps(model, V, T, z)
+    μ̄² = model.params.dipole2.values
+    Q̄² = model.params.quadrupole2.values
+    dipole_comps = []
+    quadrupole_comps = []
+    for i in @comps
+        if !iszero(μ̄²[i]) push!(dipole_comps,i) end
+        if !iszero(Q̄²[i]) push!(quadrupole_comps,i) end
+    end
+    return dipole_comps, quadrupole_comps
+end
+
+
+function J2(model, V, T, z, type, i, j, η = @f(ζ,3),m = model.params.segment.values,ϵ = model.params.epsilon.values)
     ϵT = ϵ[i,j]/T
-    aij = _ppcsaft_corr_poly(corr_a,m̄)
-    bij = _ppcsaft_corr_poly(corr_b,m̄)
+    m̄ = sqrt(m[i]*m[j])
+    if type == "DD" m̄ = min(m̄,2*one(η)) end
+    consts = eval(Symbol(type*"_consts"))
+    aij = _ppcsaft_corr_poly(consts.corr_a,m̄)
+    bij = _ppcsaft_corr_poly(consts.corr_b,m̄)
     cij = aij .+ bij .* ϵT
     return evalpoly(η,cij)
 end
 
-function J3(model, V, T, z, i, j, k, η = @f(ζ,3),m = model.params.segment.values)
-    corr_c = PPCSAFTconsts.corr_c
-    m̄ = min(cbrt(m[i]*m[j]*m[k]),2*one(η))
+function J3(model, V, T, z, type, i, j, k, η = @f(ζ,3),m = model.params.segment.values)
+    corr_c = eval(Symbol(type*"_consts")).corr_c

```

Comment:
try to write this in a conditional instead, this is really bad in terms of performance. you could also do something like:
```
function get_ppcsaft_consts(x)
if x == :DD
  return DD_consts
elseif x ==:QQ
  return QQ_consts
else
  return DQ_consts #fallback, we always return something of the same type
end
end


```